File: frontend/src/api/crm.ts
import { buildQuery, downloadFile, request } from "./client";

export interface CrmCategory { id: string; parent_id: string | null; label: string; hint: string; rules: string[]; disabled: boolean; custom: boolean }
export interface CrmInstruction { key: string; title: string; body: string; steps: string[]; revision: number }
export interface CrmCatalog { parks: string[]; cities: string[]; categories: CrmCategory[]; instructions: Record<string, CrmInstruction> }
export interface CrmInput {
  request_id: string; channel: string; phone: string; license_number: string; driver_id: string;
  contacted_at: string; park: string; city: string; category_ids: string[]; details: Record<string, string>;
  comment: string; is_ticket: boolean;
}
export interface CrmAppeal extends Omit<CrmInput, "request_id"> {
  id: number; author_id: number; author_name: string; category_labels: string[];
  status: string; created_at: string; attachments: { id: number; name: string; mime: string; size: number }[];
}
export const CRM_STATUS: Record<string, string> = { recorded: "Зафиксировано", new: "Новый тикет", in_progress: "В работе", closed: "Закрыт" };
export const CRM_FIELDS: Record<string, string> = { company: "Название компании", callback: "Номер для обратного звонка", service: "Предлагаемая услуга", employee: "Имя сотрудника", transaction: "Транзакция / РРН", conditions: "Условия работы", error_description: "Описание ошибки" };
export const crm = {
  catalog: () => request<CrmCatalog>("/api/v1/learning/crm/catalog"),
  instruction: ({ key, ...input }: CrmInstruction) => request<CrmInstruction>(`/api/v1/admin/learning/crm/instructions/${encodeURIComponent(key)}`, { method: "PUT", json: input }),
  list: (params: Record<string, unknown>) => request<{ items: CrmAppeal[]; total: number; page: number; size: number; counts: Record<string, number> }>(`/api/v1/learning/crm/appeals${buildQuery(params)}`),
  appeal: (id: number) => request<CrmAppeal>(`/api/v1/learning/crm/appeals/${id}`),
  create: (input: CrmInput, files: File[]) => {
    const multipart = new FormData(); multipart.append("payload", JSON.stringify(input));
    files.forEach(file => multipart.append("files", file));
    return request<CrmAppeal>("/api/v1/learning/crm/appeals", { method: "POST", multipart });
  },
  category: (input: { label: string; parent_id: string | null; hint: string }) => request<CrmCategory>("/api/v1/admin/learning/crm/categories", { method: "POST", json: input }),
  status: (id: number, status: string) => request<CrmAppeal>(`/api/v1/admin/learning/crm/appeals/${id}/status`, { method: "PATCH", json: { status } }),
  download: (id: number, name: string) => downloadFile(`/api/v1/learning/crm/attachments/${id}`, name),
};

export function selectedCategories(nodes: CrmCategory[], ids: string[]) { return ids.map(id => nodes.find(node => node.id === id)).filter((node): node is CrmCategory => !!node); }
export function categoryRules(nodes: CrmCategory[], ids: string[]) { return new Set(selectedCategories(nodes, ids).flatMap(node => node.rules)); }
export function extraFields(rules: Set<string>) {
  return [...(rules.has("cooperation") ? ["company", "callback", "service"] : []), ...(rules.has("employee") ? ["employee"] : []), ...(rules.has("transaction") ? ["transaction"] : []), ...(rules.has("conditions") ? ["conditions"] : []), ...(rules.has("error_description") ? ["error_description"] : [])];
}

File: app/schemas/crm.py
from datetime import UTC, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AppealInput(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    request_id: UUID
    channel: Literal["Звонок", "Чат"] = "Звонок"
    phone: str = Field(min_length=5, max_length=40)
    license_number: str = Field(min_length=1, max_length=80)
    driver_id: str = Field(default="", max_length=80)
    contacted_at: datetime
    park: str = Field(max_length=100)
    city: str = Field(max_length=100)
    category_ids: list[str] = Field(min_length=1, max_length=10)
    details: dict[str, str] = Field(default_factory=dict, max_length=10)
    comment: str = Field(default="", max_length=10000)
    is_ticket: bool = False

    @field_validator("contacted_at")
    @classmethod
    def timezone_required(cls, value):
        if value.tzinfo is None:
            raise ValueError("Дата обращения должна содержать часовой пояс")
        return value.astimezone(UTC)

    @field_validator("details")
    @classmethod
    def bound_details(cls, value):
        allowed = {
            "company",
            "callback",
            "service",
            "employee",
            "transaction",
            "conditions",
            "error_description",
        }
        if set(value) - allowed or any(len(v) > 2000 for v in value.values()):
            raise ValueError("Некорректные дополнительные поля")
        return {k: v.strip() for k, v in value.items()}


class CategoryInput(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    parent_id: str | None = Field(default=None, max_length=64)
    label: str = Field(min_length=1, max_length=160)
    hint: str = Field(default="", max_length=2000)


class StatusInput(BaseModel):
    status: Literal["new", "in_progress", "closed"]
